### ISSUE TITLE
FirebaseSDK increase pre-main app launch time

### ISSUE THREAD

**Issue by dangthaison91** [firebase/firebase-ios-sdk#5956]
### [REQUIRED] Step 1: Describe your environment

  * Xcode version:11.5
  * Firebase SDK version: 6.26.0
  * Firebase Component: Auth, Core, Analytics, Firestore, Messaging, RemoteConfig, Crashlytics, Performance...
  * Component version: 6.26.0
  * Installation method: Manual install (Bazel) base on CocoaPods

### [REQUIRED] Step 2: Describe the problem

We are optimizing our **app launch time** cause it will improve user experience. We found that using Firebase increase **pre-main** time so much. `pre-main` is time need to initialize App before App `main` and `willFinishLaunching` being called.

```ruby
Total pre-main time: 682.41 milliseconds (100.0%)
         dylib loading time:  41.98 milliseconds (6.1%)
        rebase/binding time:  39.97 milliseconds (5.8%)
            ObjC setup time: 255.26 milliseconds (37.4%)
           initializer time: 345.18 milliseconds (50.5%)
           slowest intializers :
             libSystem.B.dylib :   7.50 milliseconds (1.1%)
   libBacktraceRecording.dylib :  13.97 milliseconds (2.0%)
    libMainThreadChecker.dylib :  50.82 milliseconds (7.4%)
                   AppCore : 472.60 milliseconds (69.2%)

  total time: 2.5 seconds (100.0%)
  total images loaded:  520 (513 from dyld shared cache)
  total segments mapped: 23, into 413 pages
  total images loading time: 1.2 seconds (49.1%)
  total load time in ObjC: 255.26 milliseconds (9.8%)
  total debugger pause time: 1.2 seconds (47.5%)
  total dtrace DOF registration time:   0.00 milliseconds (0.0%)
  total rebase fixups:  167,797
  total rebase fixups time:  38.82 milliseconds (1.5%)
  total binding fixups: 732,119
  total binding fixups time: 668.10 milliseconds (25.8%)
  total weak binding fixups time:   4.80 milliseconds (0.1%)
  total redo shared cached bindings time: 671.76 milliseconds (26.0%)
  total bindings lazily fixed up: 0 of 0
  total time in initializers and ObjC +load: 345.18 milliseconds (13.3%)
                         libSystem.B.dylib :   7.50 milliseconds (0.2%)
               libBacktraceRecording.dylib :  13.97 milliseconds (0.5%)
                           libobjc.A.dylib :   2.60 milliseconds (0.1%)
                libMainThreadChecker.dylib :  50.82 milliseconds (1.9%)
              libViewDebuggerSupport.dylib :   6.71 milliseconds (0.2%)
                               AppCore : 472.60 milliseconds (18.3%)
total symbol trie searches:    1687654
total symbol table binary searches:    0
total images defining weak symbols:  53
total images using weak symbols:  129
```

As you can see, ` total binding fixups time` and `total binding fixups time`  take a bunch of time.

We check if Firebase SDKs using deprecated API `[Objective-C +load ]` by adding break-point at the end of AppDelegate didFinishLaunching, and run follow command in Xcode console:

```bash
br s -r "\+\[.+ load\]$"
br list
```
Results:

```bash
  2.48: where = AppCore`+[FPRAppActivityTracker load], address = 0x0000000112721581, resolved, hit count = 0 
  2.49: where = AppCore`+[FPRClient load], address = 0x000000011272849d, resolved, hit count = 0 
  2.50: where = AppCore`+[FPRCCLogger load], address = 0x0000000112739b91, resolved, hit count = 0 
  2.51: where = AppCore`+[FIRAnalyticsConnector load], address = 0x000000011278abab, resolved, hit count = 0 
  2.52: where = AppCore`+[FIRExperimentController load], address = 0x00000001127929f8, resolved, hit count = 0 
  2.53: where = AppCore`+[FIRMessaging load], address = 0x00000001127945be, resolved, hit count = 0 
  2.54: where = AppCore`+[FIRIAMDefaultDisplayImpl load], address = 0x00000001127be5f5, resolved, hit count = 0 
  2.55: where = AppCore`+[FIRInAppMessaging load], address = 0x00000001127d3221, resolved, hit count = 0 
  2.56: where = AppCore`+[FIRDynamicLinks load], address = 0x00000001128243a0, resolved, hit count = 0 
  2.57: where = AppCore`+[FIRInstanceID load], address = 0x0000000112829d1e, resolved, hit count = 0 
  2.58: where = AppCore`+[FIRMLKitAnalyticsLogger load], address = 0x000000011283af4d, resolved, hit count = 0 
  2.59: where = AppCore`+[FIRVision load], address = 0x000000011289c1ab, resolved, hit count = 0 
  2.60: where = AppCore`+[GMVLogger load], address = 0x00000001128b1126, resolved, hit count = 0 
  2.61: where = AppCore`+[FIRRemoteConfigComponent load], address = 0x0000000112f5647c, resolved, hit count = 0 
  2.62: where = AppCore`+[APMAnalytics load], address = 0x0000000112f838da, resolved, hit count = 0 
  2.63: where = AppCore`+[APMMeasurement load], address = 0x0000000112f9eda3, resolved, hit count = 0 
  2.64: where = AppCore`+[UIViewController(APMScreenClassName) load], address = 0x0000000112fedc29, resolved, hit count = 0 
  2.65: where = AppCore`+[GTMSessionFetcher load], address = 0x00000001130103e0, resolved, hit count = 0 
  2.66: where = AppCore`+[GTMSessionUploadFetcher load], address = 0x0000000113022570, resolved, hit count = 0 
  2.67: where = AppCore`+[FSTFirestoreComponent load], address = 0x000000011315a508, resolved, hit count = 0 
  2.68: where = AppCore`+[FIRAuth load], address = 0x0000000113348df4, resolved, hit count = 0 
  2.69: where = AppCore`+[GDTCORFlatFileStorage load], address = 0x000000011337e33b, resolved, hit count = 0 
  2.70: where = AppCore`+[GDTCORLifecycle load], address = 0x0000000113380618, resolved, hit count = 0 
  2.71: where = AppCore`+[GDTCORApplication load], address = 0x0000000113381a2d, resolved, hit count = 0 
  2.72: where = AppCore`+[GDTCCTPrioritizer load], address = 0x0000000113388075, resolved, hit count = 0 
  2.73: where = AppCore`+[GDTCCTUploader load], address = 0x00000001133899b7, resolved, hit count = 0 
  2.74: where = AppCore`+[FIRInstallations load], address = 0x00000001133b0e50, resolved, hit count = 0 
  2.75: where = AppCore`+[FIRCrashlytics load], address = 0x00000001133ef890, resolved, hit count = 0
```

Could you please replace **+load** with **+initialise** function? Any improvement could help to reduce **binding fixups time**?

#### Steps to reproduce:

1. Create an empty project
2. Add Firebase SDK, import then add `configureApp` function
3. Product -> Scheme -> Edit Scheme -> Add 2 build arguments: **DYLD_PRINT_STATISTICS** and **DYLD_PRINT_STATISTICS_DETAILS**
4. Build with Release optimization
5. Check LLVM debug messages are printed out on Xcode console

**- Project build and measure on iPhone 6S (reboot, no background/open apps)**

### References:
- [Reducing Your App's Launch Time](https://developer.apple.com/documentation/xcode/improving_your_app_s_performance/reducing_your_app_s_launch_time)
- https://developer.apple.com/videos/play/wwdc2016/406/
- https://useyourloaf.com/blog/slow-app-startup-times/

**Comment by dangthaison91** [comment]
`pre-main` time for an empty app:

```ruby
Total pre-main time:  89.70 milliseconds (100.0%)
         dylib loading time:  27.23 milliseconds (30.3%)
        rebase/binding time:   4.65 milliseconds (5.1%)
            ObjC setup time:  15.19 milliseconds (16.9%)
           initializer time:  42.61 milliseconds (47.5%)
           slowest intializers :
             libSystem.B.dylib :   3.57 milliseconds (3.9%)
   libBacktraceRecording.dylib :   6.98 milliseconds (7.7%)
    libMainThreadChecker.dylib :  20.42 milliseconds (22.7%)
  libViewDebuggerSupport.dylib :   7.13 milliseconds (7.9%)

  total time: 783.05 milliseconds (100.0%)
  total images loaded:  396 (390 from dyld shared cache)
  total segments mapped: 21, into 400 pages
  total images loading time: 496.66 milliseconds (63.4%)
  total load time in ObjC:  15.19 milliseconds (1.9%)
  total debugger pause time: 469.43 milliseconds (59.9%)
  total dtrace DOF registration time:   0.00 milliseconds (0.0%)
  total rebase fixups:  18,064
  total rebase fixups time:  17.35 milliseconds (2.2%)
  total binding fixups: 528,059
  total binding fixups time: 211.17 milliseconds (26.9%)
  total weak binding fixups time:   0.03 milliseconds (0.0%)
  total redo shared cached bindings time: 223.91 milliseconds (28.5%)
  total bindings lazily fixed up: 0 of 0
  total time in initializers and ObjC +load:  42.61 milliseconds (5.4%)
                         libSystem.B.dylib :   3.57 milliseconds (0.4%)
               libBacktraceRecording.dylib :   6.98 milliseconds (0.8%)
                           libobjc.A.dylib :   1.34 milliseconds (0.1%)
                libMainThreadChecker.dylib :  20.42 milliseconds (2.6%)
              libViewDebuggerSupport.dylib :   7.13 milliseconds (0.9%)
total symbol trie searches:    1320558
total symbol table binary searches:    0
total images defining weak symbols:  41
total images using weak symbols:  106

```

**Comment by paulb777** [comment]
@dangthaison91 Thanks for the report.  A good next step would be to try to isolate the startup slowness to specific Firebase pods and then use Instruments profiling to go deeper.

**Comment by sgammon** [comment]
@dangthaison91 how did you get Bazel to work with Firebase on iOS? Did you use PodToBuild, or a different integration method?

**Comment by dangthaison91** [comment]
@sgammon We do not rebuild Firebase with Bazel, instead we use pre-built frameworks from Firebase releases.

**Comment by sgammon** [comment]
@dangthaison91 understood, but with `.xcframework`s?

**Comment by dangthaison91** [comment]
We supported `.xcframeworks` but not migrate yet.

**Comment by dangthaison91** [comment]
It has been about 2 months, any updates on this issue @paulb777 We still looking to see good news from Firebase teams.

**Comment by paulb777** [comment]
@dangthaison91 Sorry about the delay. We haven't been able to find the bandwidth to dig too deep into this yet. Two things that would help:

- A sample app that demonstrates the `pre-main` times documented above
- Isolating the slow pre-main time to specific Firebase libraries

**Comment by jserrafindster** [comment]
Not sure if might be related but,

i'm also having an issue happening randomly with launch time sometimes taking over 10 to launch. I tried to analyze it and found it coming from Firebase's configureapp. I was able then to pause the code when this happen to see where the code was running. 

I dont have it in text but heres the screenshot
![Screenshot 2020-09-03 at 16 57 24](https://user-images.githubusercontent.com/52421199/92138728-ab565e00-ee06-11ea-82f8-808cdb8218e7.png)

Later I also found that we had 180 pending blocks
![Screenshot 2020-09-03 at 16 59 38](https://user-images.githubusercontent.com/52421199/92138919-efe1f980-ee06-11ea-8ff4-f89f048ad9c4.png)

**Comment by maksymmalyhin** [comment]
@jserrafindster It looks like a separate issue. Could you please create a separate ticket for it? From the first glance it looks like an issue with an excessive amount of requests sent by Firebase Installations in the case of Firebase miss-configuration which was fixed in #6232. If it looks different, let's continue the discussion in a separate ticket.

**Comment by ryanwilson** [comment]
Hi @dangthaison91,

We still need to audit the `+load` calls, although I do know off the top of my head that some are required in order to catch the right lifecycle events needed for Firebase to function correctly (ex. we have a components system that is used for weak-dependencies that needs to happen before any calls to those classes are made). If there are specific functions you see are causing startup latency (for example, issues #6902 and #2669) that would be more actionable. Thanks again!

**Comment by ryanwilson** [comment]
Closing this now since there have been some improvements lately and there's not specific action we can take to close this bug, if there are any specific functions causing startup latency that are more actionable please file a new issue and we'd be happy to take a look. Thanks for the report!
